src/core/socket.h

```diff
@@ -162,11 +167,18 @@ struct Socket {
         RateLimit trigger_limit;
 };
 
+typedef struct SocketPeer {
+        unsigned n_connections;
+
+        int family;
+        union sockaddr_union peer;
+} SocketPeer;
```

Comment:
So, I you kinda already implement a reference counted object here, as each SocketPeer object is kept alive by the Service objects that share the same peer. I think we should make this explicit, and follow the usual logic for reference counted objects on this one too. Specifically, I think that "n_connections" should be renamed "n_ref" (since that's how we usually call the reference counter), and then each Service object of the same peer maintains a direct reference to it, i.e. gains a new `SocketPeer* peer` field directly in the Service structure.

That way the connections_by_address and connections_by_fd data structures in the Socket object can be replaced by a single "peers_by_address" hash table. And SocketPeer of course needs a ptr back to the Scoket.

I hope this makes sense?